### PR TITLE
Westinm/trimming

### PR DESCRIPTION
**Pull request by westin-m**
Make IdWeb trimmable. Where possible, change implementations to be trim-friendly. Otherwise, bubble-up the warnings to Apps calling our library.  More info on trimming is [here](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-self-contained).

### Published File Size in KB Using the [4-1-MyOrg Sample](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/4-WebApp-your-API/4-1-MyOrg) (.net7)
| Filename | Before | After | % Change |
|-|-|-|-|
Microsoft.Identity.Web.Certificate.dll | 28 | 22 | 21%
Microsoft.Identity.Web.Certificateless.dll | 12 | 12 | 0%
Microsoft.Identity.Web.Diagnostics.dll | 13.5 | 9 | 33%
Microsoft.Identity.Web.dll | 149 | 149 | 0%
Microsoft.Identity.Web.DownstreamApi.dll | 53.5 | 53.5 | 0%
Microsoft.Identity.Web.TokenAcquisition.dll | 122 | 88.5 | 27%
Microsoft.Identity.Web.TokenCache.dll | 46 | 18.5 | 60%
Microsoft.Identity.Web.UI.dll | 15 | 15 | 0%
Microsoft.Identity.Web.UI.Views.dll | 17.5 | 17.5 | 0%

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -29,6 +29,12 @@
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0'">
+    <EnableTrimAnalyzer>true</EnableTrimAnalyzer>
+    <PublishTrimmed>true</PublishTrimmed>
```
Comment (bgavrilMS) [inline Directory.Build.props:34]: Is this needed for libraries? I guess it's not a problem to have it there.

File: Directory.Build.props [hunk 2]
```diff
@@ -29,6 +29,12 @@
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0'">
```
Comment (bgavrilMS) [inline Directory.Build.props:32]: @westin-m - how are you meansuring the trimmed size? Do you just look at the size of the DLL ?

File: Directory.Build.props (hunk 1 above)
Comment (jmprieur) [reply Directory.Build.props:34]: Those are libraries, only the exec should be trimmed.

File: Directory.Build.props (hunk 2 above)
Comment (westin-m) [reply Directory.Build.props:32]: Currently yes, I am looking at the size of the .dll files.

File: Directory.Build.props (hunk 1 above)
Comment (westin-m) [reply Directory.Build.props:34]: > Is this needed for libraries? I guess it's not a problem to have it there.

This is not needed and will be replaced with `IsTrimmable`.

File: src/Microsoft.Identity.Web/DownstreamWebApiSupport/DownstreamWebApiGenericExtensions.cs [hunk 3]
```diff
@@ -276,12 +295,16 @@ public static class DownstreamWebApiGenericExtensions
 
             return await ConvertToOutput<TOutput>(response).ConfigureAwait(false);
         }
-
+#if NET6_0_OR_GREATER
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/DownstreamWebApiSupport/DownstreamWebApiGenericExtensions.cs:298]: nit add a space after the }

File: src/Microsoft.Identity.Web/DownstreamWebApiSupport/DownstreamWebApiGenericExtensions.cs [hunk 4]
```diff
@@ -276,12 +295,16 @@ public static class DownstreamWebApiGenericExtensions
...
+        [RequiresUnreferencedCode("Calls System.Text.Json.JsonSerializer.Serialize<TValue>(TValue, JsonSerializerOptions).")]
+#endif
         private static StringContent ConvertFromInput<TInput>(TInput input)
         {
             return new StringContent(JsonSerializer.Serialize(input), Encoding.UTF8, "application/json");
         }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/DownstreamWebApiSupport/DownstreamWebApiGenericExtensions.cs:304]: nit: add a space

File: src/Microsoft.Identity.Web/Policy/ScopeAuthorizationHandler.cs [hunk 5]
```diff
@@ -35,6 +36,7 @@ public ScopeAuthorizationHandler(IConfiguration configuration)
         /// <param name="context">AuthorizationHandlerContext.</param>
         /// <param name="requirement">Scope authorization requirement.</param>
         /// <returns>Task.</returns>
+
         protected override Task HandleRequirementAsync(
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/Policy/ScopeAuthorizationHandler.cs:40]: remove line?

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: src/Microsoft.Identity.Web.TokenAcquisition/MicrosoftIdentityBaseAuthenticationBuilder.cs [hunk 6]
```diff
@@ -22,6 +23,9 @@ public class MicrosoftIdentityBaseAuthenticationBuilder
         /// </summary>
         /// <param name="services">The services being configured.</param>
         /// <param name="configurationSection">Optional configuration section.</param>
+#if NET6_0_OR_GREATER
+        [RequiresUnreferencedCode("Calls Microsoft.Extensions.Configuration.ConfigurationBinder.Bind(IConfiguration, Object).")]
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.TokenAcquisition/MicrosoftIdentityBaseAuthenticationBuilder.cs:27]: Can we do better in .NET 8? by enabling `EnableMicrosoftExtensionsConfigurationBinderSourceGenerator` ?
https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#source-generator-for-configuration-binding

File: Directory.Build.props [hunk 7]
```diff
@@ -29,6 +29,10 @@
...
+    <IsTrimmable>true</IsTrimmable>
+  </PropertyGroup>
```
Comment (jmprieur) [inline Directory.Build.props:34]: Do we also want to try 
```xml
<EnableMicrosoftExtensionsConfigurationBinderSourceGenerator>true</EnableMicrosoftExtensionsConfigurationBinderSourceGenerator>
```
if we use [Microsoft.Extensions.Configuration.Binder NuGet package](https://www.nuget.org/packages/Microsoft.Extensions.Configuration.Binder) 8.*?

and then remove the attributes which are due to the binding.

Shall we do that when we enable .NET 8?
@jennyf19 @westin-m ?

**Review by jmprieur — APPROVED** [review]
LGTM
I left a question

We can address it in a later iteration (with .NET 8). But would want to capture that in a work item.

File: Directory.Build.props (hunk 7 above)
Comment (jennyf19) [reply Directory.Build.props:34]: Let's do that as an increment.

Comment (westin-m) [reply Directory.Build.props:34]: Yes, we should do this with .NET 8.

File: src/Microsoft.Identity.Web.TokenAcquisition/MicrosoftIdentityBaseAuthenticationBuilder.cs (hunk 6 above)
Comment (westin-m) [reply src/Microsoft.Identity.Web.TokenAcquisition/MicrosoftIdentityBaseAuthenticationBuilder.cs:27]: From what I read, probably. I do want to try this and test it when we add .NET 8.

File: Directory.Build.props (hunk 7 above)
Comment (westin-m) [reply Directory.Build.props:34]: Yes, I want to try this with .NET 8.

**Pull request merged**